gateway-js/src/config.ts

```diff
@@ -137,7 +137,9 @@ interface GatewayConfigBase {
 
 export interface RemoteGatewayConfig extends GatewayConfigBase {
   serviceList: ServiceEndpointDefinition[];
-  introspectionHeaders?: HeadersInit;
+  introspectionHeaders?:
+    | HeadersInit
+    | ((service: ServiceEndpointDefinition) => Promise<HeadersInit>);
```

Comment:
We can make the user provided fn optionally `async` by changing the return type:
```suggestion
    | ((service: ServiceEndpointDefinition) => Promise<HeadersInit> | HeadersInit);
```
And since it's fine to `await` a non-async function, we don't have to change anything about how we're calling and `await`ing that function now.

Reply:
:+1: I wanted to do this from the get-go but was worried about how to properly deal with it, but your comment [here](https://github.com/apollographql/federation/pull/607#discussion_r603523662) answers it :)